[1.1.x] Update to new SdFat Lib

unimatrix099 opened the pull request:
Updated the code to use the SdFat library from the Arduino repository.
If `USE_NEW_SD_FAT_LIB` is not defined then the old library is used.

Had to include "Arduino.h" in `MarlinConfig.h` to compile it and had to rename some pins from FastIO which where conflicting with Arduino names. Any idea on a more elegant solution?

On Trigorilla board using the new library and the bench test form examples I got ~600k/s speed, so the hardware has some raw speed there... :) {also `SdFatEX` was enabled, the `ENABLE_EXTENDED_TRANSFER_CLASS` set to 1 in `SdFatConfig.h` in SdFat library}

thinkyhead commented:
Well done! I believe the only other thing is to make sure the SDFat code that we include with Marlin is not compiled or linked into the binary when using the external library. GCC may be smart enough to strip any functions that are never called, but we like to make things like this explicit when possible. I'm not currently sure which of our `Sd*.cpp`/`Sd*.h` files are supplanted by the external library, but at the top of the files that _are_, we should make this change…
```diff
- #if ENABLED(SDSUPPORT)
+ #if ENABLED(SDSUPPORT) && DISABLED(USE_NEW_SD_FAT_LIB)
```

thinkyhead commented:
> Had to include "Arduino.h" in MarlinConfig.h

We should look closer at that. I don't think a second include of `Arduino.h` should be needed in `MarlinConfig.h`. What errors do you get without it? (A small sample will do.)

> rename some pins from FastIO … conflicting with Arduino names

I'm not sure why that's happening. What version of Arduino IDE are you using?

unimatrix099 commented:
Thank you! 
Good idea to add && DISABLED(USE_NEW_SD_FAT_LIB) in Sd* files. Only the #include "SdFile.h" is the top include for the old SdFat library, the same is also for the new one, but it's SdFat.h So it should not be any conflicts. 

About the "Arduino.h" include issue, I moved the development to my pc and installed Arduino 1.8.5, no more issues, checked out the branch with the new modifications on the laptop Arduino 1.8.2, no more issues. It seems the cleanup for the PR fixed the issue :)) So, I reverted the fastio renaming of the pins. The issue was that SS was defined in Arduino and also in fast IO. But in the end it's not an issue anymore.

unimatrix099 commented:
I'll take that back... I forgot SDCARD disabled... please don't merge this PR until it's fixed.

unimatrix099 committed:
updated to new SdFat Lib

unimatrix099 commented:
The issue with Arduino.h is that when the new SdFat lib is enabled it complains about not having Print and Serial:
```

libraries/SdFat/src/FatLib/FatVolume.h:53:9: error: 'Print' does not name a type
 typedef Print print_t;
         ^
libraries/SdFat/src/FatLib/FatFile.h:146:9: error: 'Serial' was not declared in this scope
     ls(&Serial, flags);
         ^

```
If I move Arduino.h to be the first include in MarlinConfig.h before the fastio.h then I get the error:
```

In file included from sketch/fastio.h:50:0,
                 from sketch/MarlinConfig.h:27,
                 from sketch/Marlin.h:36,
                 from sketch/Marlin_main.cpp:250:
fastio_1280.h:47: error: 'DIO53' was not declared in this scope
 #define SS          DIO53
                     ^
/home/adrian/Arduino/libraries/SdFat/src/SdFat.h:326:30: note: in expansion of macro 'SS'
   bool begin(uint8_t csPin = SS, SPISettings spiSettings = SPI_FULL_SPEED) {
                              ^
fastio_1280.h:47: error: 'DIO53' was not declared in this scope
 #define SS          DIO53
                     ^
/home/adrian/Arduino/libraries/SdFat/src/SdFat.h:336:34: note: in expansion of macro 'SS'
   bool cardBegin(uint8_t csPin = SS, SPISettings settings = SPI_FULL_SPEED) {
                                  ^
fastio_1280.h:47: error: 'DIO53' was not declared in this scope
 #define SS          DIO53
                     ^
/home/adrian/Arduino/libraries/SdFat/src/SdFat.h:459:30: note: in expansion of macro 'SS'
   bool begin(uint8_t csPin = SS, SPISettings spiSettings = SPI_FULL_SPEED) {
                              ^
fastio_1280.h:47: error: 'DIO53' was not declared in this scope
 #define SS          DIO53
                     ^
/home/adrian/Arduino/libraries/SdFat/src/SdFat.h:503:30: note: in expansion of macro 'SS'
   bool begin(uint8_t csPin = SS, SPISettings settings = SD_SCK_MHZ(50)) {
                              ^
exit status 1
'DIO53' was not declared in this scope

```

If I comment  `//#define SS          DIO53` from fastio then it compiles...

Any ideas on how to fix this in an elegant way?

unimatrix099 commented:
In Arduino/hardware/arduino/avr/variants/mega/pins_arduino.h  the SS variable is defined... 

```
#define PIN_SPI_SS    (53)
#define PIN_SPI_MOSI  (51)
#define PIN_SPI_MISO  (50)
#define PIN_SPI_SCK   (52)

static const uint8_t SS   = PIN_SPI_SS;
static const uint8_t MOSI = PIN_SPI_MOSI;
static const uint8_t MISO = PIN_SPI_MISO;
static const uint8_t SCK  = PIN_SPI_SCK;
```

So... how should we proceed?

thinkyhead commented:
This kind of change won't be getting made to 1.1.x, as the 1.1 version is soon to get retired. Please submit to 2.0.x if this still seems like a viable concept.
